src/shogun/multiclass/tree/CARTree.cpp

```diff
@@ -532,17 +571,62 @@ int32_t CCARTree::compute_best_attribute(const SGMatrix<float64_t>& mat, const S
 	float64_t max_gain=MIN_SPLIT_GAIN;
 	int32_t best_attribute=-1;
 	float64_t best_threshold=0;
+
+	SGVector<int32_t> indices_mask;
+	SGVector<int32_t> count_indices(mat.num_rows);
+	count_indices.zero();
+	SGVector<int32_t> dupes(num_vecs);
+	dupes.range_fill();
+	if (m_pre_sort)
+	{
+		indices_mask = SGVector<int32_t>(mat.num_rows);
+		indices_mask.set_const(-1);
+		for(int32_t j=0;j<active_indices.size();j++)
+		{
+			if (indices_mask[active_indices[j]]>=0)
+				dupes[indices_mask[active_indices[j]]]=j;
+			indices_mask[active_indices[j]]=j;
+			count_indices[active_indices[j]]++;
+		}
+	}
+
 	for (int32_t i=0;i<num_feats;i++)
 	{
 		SGVector<float64_t> feats(num_vecs);
-		for (int32_t j=0;j<num_vecs;j++)
-			feats[j]=mat(idx[i],j);
-
-		// O(N*logN)
 		SGVector<index_t> sorted_args(feats.size());
-		sorted_args.range_fill();
-		CMath::qsort_index(feats.vector, sorted_args.vector, feats.size());
+		SGVector<int32_t> temp_count_indices(count_indices.size());
+		memcpy(temp_count_indices.vector, count_indices.vector, sizeof(int32_t)*count_indices.size());
 
+		if (m_pre_sort)
+		{
+			SGVector<float64_t> temp_vec(mat.get_column_vector(idx[i]), mat.num_rows, false);
+			SGVector<int32_t> sorted_indices(m_sorted_indices.get_column_vector(idx[i]), mat.num_rows, false);
+			int32_t count=0;
+			for(int32_t j=0;j<mat.num_rows;j++)
```

Comment:
wait, why are you still traversing row wise?
This outerloop should be over columns